packages/ses/src/module-shim.js

```diff
@@ -0,0 +1,237 @@
+defineProperties(StaticModuleRecord, {
+  prototype: { value: StaticModuleRecordPrototype },
+});
+
+defineProperties(InertStaticModuleRecord, {
+  prototype: { value: StaticModuleRecordPrototype },
+});
+
+// moduleAliases associates every public module exports namespace with its
+// corresponding compartment and specifier so they can be used to link modules
+// across compartments.
+// The mechanism to thread an alias is to use the compartment.module function
+// to obtain the exports namespace of a foreign module and pass it into another
+// compartment's moduleMap constructor option.
+const moduleAliases = new WeakMap();
+
+// privateFields captures the private state for each compartment.
+const privateFields = new WeakMap();
+
+// Compartments do not need an importHook or resolveHook to be useful
+// as a vessel for evaluating programs.
+// However, any method that operates the module system will throw an exception
+// if these hooks are not available.
+const assertModuleHooks = compartment => {
+  const { importHook, resolveHook } = privateFields.get(compartment);
+  if (typeof importHook !== 'function' || typeof resolveHook !== 'function') {
+    throw new TypeError(
+      `Compartment must be constructed with an importHook and a resolveHook for it to be able to load modules`,
+    );
+  }
+};
+
+const ModularCompartmentPrototypeExtension = {
+  constructor: InertModularCompartment,
+
+  module(specifier) {
+    if (typeof specifier !== 'string') {
+      throw new TypeError('first argument of module() must be a string');
+    }
+
+    assertModuleHooks(this);
+
+    const { exportsProxy } = getDeferredExports(
+      this,
+      privateFields.get(this),
+      moduleAliases,
+      specifier,
+    );
+
+    return exportsProxy;
+  },
+
+  async import(specifier) {
+    if (typeof specifier !== 'string') {
+      throw new TypeError('first argument of import() must be a string');
+    }
+
+    assertModuleHooks(this);
+
+    return load(privateFields, moduleAliases, this, specifier).then(() => {
+      const namespace = this.importNow(specifier);
+      return { namespace };
+    });
+  },
+
+  async load(specifier) {
+    if (typeof specifier !== 'string') {
+      throw new TypeError('first argument of load() must be a string');
+    }
+
+    assertModuleHooks(this);
+
+    return load(privateFields, moduleAliases, this, specifier);
+  },
+
+  importNow(specifier) {
+    if (typeof specifier !== 'string') {
+      throw new TypeError('first argument of importNow() must be a string');
+    }
+
+    assertModuleHooks(this);
+
+    const moduleInstance = link(
+      privateFields,
+      moduleAnalyses,
+      moduleAliases,
+      this,
+      specifier,
+    );
+    moduleInstance.execute();
+    return moduleInstance.exportsProxy;
+  },
+};
+
+const ModularCompartmentPrototype = create(Object.prototype, {
+  ...getOwnPropertyDescriptors(CompartmentPrototype),
+  ...getOwnPropertyDescriptors(ModularCompartmentPrototypeExtension),
+});
```

Comment:
I'd like to mention a suggestion I have mixed feeling about. Altogether, I think I like the status quo better than my suggestion, but I mention it so we can talk about it.

```suggestion
defineProperties(
  CompartmentPrototype, 
  getOwnPropertyDescriptors(ModularCompartmentPrototypeExtension),
);
```

Rather than creating a new `ModularCompartmentPrototype`, just add the new methods to the one `CompartmentPrototype`. In the world where there is a `ModularCompartmentPrototype`, the original `CompartmentPrototype` should no longer be visible to user code. Since we're doing all this before lockdown, we can monkey patch the `CompartmentPrototype` in this way if we wish.

***OOPS:*** In writing this review comment and looking back at the whitelist, I see I misread it in ways that make the particulars of my previous criticism at https://github.com/Agoric/SES-shim/pull/443#discussion_r479714684 invalid. The whitelist does show `%InertModularCompartment%.prototype` as `%ModularCompartmentPrototype%`. The two inert constructors to not share one prototype. Each prototype does point back at the appropriate inert constructor. Without the further questionable suggestion I make in this note, all that is perhaps as it should be. Or at least it is not as far off.

But, if we do reuse one `CompartmentPrototype` by adding the module-level messages to it, then we should indeed consolidate the inert constructors as I had previously bogusly suggested. More important, everywhere where this PR is careful to thread the correct `compartmentPrototype`, we could simplify since there is only one, independent of the level of abstraction we're operating at.